### ISSUE TITLE
Try to run temporal on top of  Amazon Keyspaces

### ISSUE THREAD

**Issue by mfateev** [temporalio/temporal#328]
They claim supporting LWT: https://docs.aws.amazon.com/keyspaces/latest/devguide/functional-differences.html

This would be great for users running on AWS.

**Comment by rfwagner** [comment]
They don't support LOGGED BATCH. Isn't that a showstopper at this point?

**Comment by samarabbas** [comment]
Yes we need ability to update multiple rows in single batch protected by condition.  So this would be a blocker.

**Comment by joebowbeer** [comment]
Some discussion here as well:

https://community.temporal.io/t/running-on-amazon-keyspaces/159/7

**Comment by robzienert** [comment]
Conditional delete is another feature missing.

**Comment by nwheeler81** [comment]
> Conditional delete is another feature missing.

I think, Amazon Keyspaces supports condition delete.
delete from table where pk=? and cl=? IF col>9;

@ Row 1
-----------+-------
 [applied] | False
 col           | 9

**Comment by freeqaz** [comment]
Any update on this, out of curiosity? I'm likely not the only person looking to answer this question. We're looking to use Temporal, but we're not sure if we should go with Cassandra or SQL.

Managing a Cassandra cluster ourselves sounds pretty bad, and I would prefer to use Keyspaces over that. If anybody is more skilled at testing out a deployment of Temporal on Keyspaces, that would be super helpful. Thanks!

**Comment by freeqaz** [comment]
Just to add some more context from poking around on existing threads:

## LWT support (added)

@mfateev said [here](https://community.temporal.io/t/aws-keyspace-as-managed-cassandra-for-temporal/1762) that they need AWS Keyspaces to add LWT support. From the time this was posted, AWS must have added this because [this](https://docs.aws.amazon.com/keyspaces/latest/devguide/functional-differences.html#functional-differences.light-transactions) page says the following:

> "Amazon Keyspaces (for Apache Cassandra) fully supports compare and set functionality on INSERT, UPDATE, and DELETE commands, which are known as lightweight transactions (LWTs) in Apache Cassandra. As a serverless offering, Amazon Keyspaces (for Apache Cassandra) provides consistent performance at any scale, including for lightweight transactions. With Amazon Keyspaces, there is no performance penalty for using lightweight transactions."

## Conditional delete support (probably added, but I'm a noob)

The comment above from @nwheeler81 shows conditional deletes working, and the Keyspaces docs say the [following](https://docs.aws.amazon.com/keyspaces/latest/devguide/functional-differences.html#functional-differences.batch):

> "Amazon Keyspaces supports unlogged batch commands with up to 30 commands in the batch. Only unconditional INSERT, UPDATE, or DELETE commands are permitted in a batch. Logged batches are not supported."

This is for batches, and nothing else is mentioned anywhere on the website. Maybe with more digging I can sort this out, but @nwheeler81's comment likely trumps the docs.

## Logged Batch support (not supported still)

This page on [supported Cassandra features](https://docs.aws.amazon.com/keyspaces/latest/devguide/cassandra-apis.html#cassandra-api-support) is probably the best place to "glance" for info.

They say that Logged Batch is unsupported, but unlogged batches are supported. I don't know enough about Temporal or Cassandra to understand what this means.

-------

That's my summary of the state of this so far. If there is anything else missing that Temporal uses, I'm happy to dig in and verify. Otherwise, it seems like the only thing missing is "Logged Batch" support.

Is there a workaround possible for that?

**Comment by ouanixi** [comment]
I think there's also a consistency level that is not matched. 
A few comments in the code state that "quorum" level consistency is required. [Here's one.](https://github.com/temporalio/temporal/blob/ad8342bd5b95f5a3c243f7d0f12ae561a91633b6/common/persistence/cassandra/cassandraPersistence.go#L2551)
Which is [not supported](https://docs.aws.amazon.com/keyspaces/latest/devguide/consistency.html) by keyspaces

**Comment by Zonalds** [comment]
hi,

Has anything changed?

**Comment by karunagoyal** [comment]
Hi all,
Can anyone clarify if it is supported?

**Comment by meet-bhagdev** [comment]
Temporal team - i am a product manager at AWS working on Amazon Keyspaces. Would love to re-engage here to understand what we can do to enable [Temporal.io](http://temporal.io/) for Amazon Keyspaces

**Comment by wojo** [comment]
@meet-bhagdev we are evaluating Temporal and would much prefer to use Keyspaces, I think the gaps are above so we'll be watching this closely.

**Comment by jamesbuddrige** [comment]
Also keen to see this! We're currently using Postgres on RDS, but would love to see keyspaces support.

**Comment by maxime-gaudron** [comment]
@meet-bhagdev any news? :)

**Comment by matanbaruch** [comment]
+1

**Comment by vrazvan-mih** [comment]
AWS Keyspaces support logged batches since November 2025. 
https://aws.amazon.com/blogs/database/amazon-keyspaces-now-supports-logged-batches-for-atomic-multi-statement-operations/

**Comment by seagyn** [comment]
Edit: I believe this was actually only for visibility which was deprecated for cassandra a while back.

I believe the missing piece is that Keypaces doesn't support create index commands which Temporal needs: https://docs.aws.amazon.com/keyspaces/latest/devguide/cassandra-apis.html#cassandra-api-support

What I understand is that you could potentially use a materialised view for a "secondary" index.
